### PR TITLE
Stage I: Pulumi abstraction

### PR DESCRIPTION
**Pull request by illusional**

### CODE REVIEW

**Comment by illusional** [comment]
In this branch, I've got all our infrastructure replicated, and matched all the pulumi IDs. There are some I've changed the resource ID because it made things a lot easier, but I've catalogued them all here. Spreadsheet that links all the old to new IDs: https://docs.google.com/spreadsheets/d/1Jv1EVvW1qKhf9oUBeCyfcRBcGhM6XWqN933xDLQmfec/edit#gid=0

**Comment by illusional** [comment]
Hey @lgruen + @violetbrina,

In this branch, I've managed to get all our infrastructure replicated, and matched all the pulumi IDs. There are some I've changed the resource ID because it made things a lot easier, but I've catalogued them all here. Spreadsheet that links all the old to new IDs: https://docs.google.com/spreadsheets/d/1Jv1EVvW1qKhf9oUBeCyfcRBcGhM6XWqN933xDLQmfec/edit#gid=0

Soon I'll:

- try deploying this to a new dataset 
- clean-up the PR + address the TODOs
- hopefully get this reviewed.

It's unfortunately a chunky review, but I've totally reimplemented our infrastructure, it's set-up well for the next stages and should work on Azure without little intervention (except Azure specific chunks). A lot of the code in this PR is just boilerplate.

Eg: instead of:

```python
access_group = create_group(...)

# it's now

@property
@lru_cache()
def access_group(self):
    return self.create_group("access")
```

This penalty is IMO very worth it given our we can organise infrastructure now, without having to pass state around to lots of variables.

**Comment by illusional** [comment]
Okay good news! Previously it was still trying to delete + recreate a few resources (like service-accounts, secrets) - with this PR (https://github.com/populationgenomics/analysis-runner/pull/460) I've managed to get all resources to match with no change (except an update to display names on a few service-accounts).

I'm getting a lot of "recreate" for memberships (BucketIAMMember, IAMMember, GroupMember, RepositoryIamMember, IamMember, SecretIamMember), but this is due to a member key change, as listed below. Although @lgruen, you mentioned this can sometimes cause a problem - the easiest may be to delete all the memberships and reapply them.

The members key selection is now standardised with this block: https://github.com/populationgenomics/cpg-infrastructure/blob/3896cd4124386ba1ba5a251cf08e473fb73a6d8f/cpg_infra/abstraction/gcp.py#L131-L147

```
~ id: "cromwell-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-full@fewgenomes.iam.gserviceaccount.com"
~ id: "cromwell-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-standard@fewgenomes.iam.gserviceaccount.com"
~ id: "cromwell-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-test@fewgenomes.iam.gserviceaccount.com"
~ id: "dataproc-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-full@fewgenomes.iam.gserviceaccount.com"
~ id: "dataproc-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-standard@fewgenomes.iam.gserviceaccount.com"
~ id: "dataproc-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-test@fewgenomes.iam.gserviceaccount.com"
~ id: "notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com" => "projects/notebooks-314505/serviceAccounts/notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com"
~ displayName: "Notebook service account for dataset fewgenomes" => "notebook-fewgenomes"
~ group: "groups/00nmf14n182x78r" => "fewgenomes-test@populationgenomics.org.au"
~ group: "groups/01d96cc00xq5nyw" => "fewgenomes-full@populationgenomics.org.au"
~ group: "groups/01pxezwc4865g59" => "fewgenomes-access@populationgenomics.org.au"
~ group: "groups/03as4poj2bbnpsf" => "fewgenomes-standard@populationgenomics.org.au"
~ group: "groups/00ihv63627t93wm" => "fewgenomes-sample-metadata-main-read@populationgenomics.org.au"
~ group: "groups/01pxezwc2ujaj2h" => "fewgenomes-sample-metadata-test-write@populationgenomics.org.au"
~ group: "groups/02nusc191j96266" => "fewgenomes-sample-metadata-test-read@populationgenomics.org.au"
~ group: "groups/02p2csry1vun1t3" => "fewgenomes-sample-metadata-main-write@populationgenomics.org.au"
~ group: "groups/02xcytpi0t1e0uw" => "fewgenomes-web-access@populationgenomics.org.au"
~ member: "group:fewgenomes-access@populationgenomics.org.au" => "fewgenomes-access@populationgenomics.org.au"
~ member: "group:fewgenomes-full@populationgenomics.org.au" => "fewgenomes-full-915@hail-295901.iam.gserviceaccount.com"
~ member: "group:fewgenomes-full@populationgenomics.org.au" => "fewgenomes-full@populationgenomics.org.au"
~ member: "group:fewgenomes-standard@populationgenomics.org.au" => "fewgenomes-standard-737@hail-295901.iam.gserviceaccount.com"
~ member: "group:fewgenomes-standard@populationgenomics.org.au" => "fewgenomes-standard@populationgenomics.org.au"
~ member: "group:fewgenomes-test@populationgenomics.org.au" => "fewgenomes-test-126@hail-295901.iam.gserviceaccount.com"
~ member: "group:fewgenomes-test@populationgenomics.org.au" => "fewgenomes-test@populationgenomics.org.au"
~ member: "serviceAccount:access-group-cache@analysis-runner.iam.gserviceaccount.com" => "access-group-cache@analysis-runner.iam.gserviceaccount.com"
~ member: "serviceAccount:analysis-runner-server@analysis-runner.iam.gserviceaccount.com" => "analysis-runner-server@analysis-runner.iam.gserviceaccount.com"
~ member: "serviceAccount:cromwell-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-full@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:cromwell-runner@cromwell-305305.iam.gserviceaccount.com" => "cromwell-runner@cromwell-305305.iam.gserviceaccount.com"
~ member: "serviceAccount:cromwell-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-standard@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:cromwell-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-test@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:dataproc-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-full@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:dataproc-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-standard@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:dataproc-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-test@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:fewgenomes-full-915@hail-295901.iam.gserviceaccount.com" => "fewgenomes-full-915@hail-295901.iam.gserviceaccount.com"
~ member: "serviceAccount:fewgenomes-standard-737@hail-295901.iam.gserviceaccount.com" => "fewgenomes-standard-737@hail-295901.iam.gserviceaccount.com"
~ member: "serviceAccount:fewgenomes-test-126@hail-295901.iam.gserviceaccount.com" => "fewgenomes-test-126@hail-295901.iam.gserviceaccount.com"
~ member: "serviceAccount:main-upload@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/main-upload@fewgenomes.iam.gserviceaccount.com"
~ member: "serviceAccount:notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com" => "projects/notebooks-314505/serviceAccounts/notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com"
~ preferredMemberKey: { ~ id: "dataproc-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-full@fewgenomes.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "cromwell-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-full@fewgenomes.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "dataproc-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-test@fewgenomes.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "cromwell-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-test@fewgenomes.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com" => "projects/notebooks-314505/serviceAccounts/notebook-fewgenomes@notebooks-314505.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "dataproc-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/dataproc-standard@fewgenomes.iam.gserviceaccount.com" }
~ preferredMemberKey: { ~ id: "cromwell-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-standard@fewgenomes.iam.gserviceaccount.com" }
~ serviceAccountId: "cromwell-full@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-full@fewgenomes.iam.gserviceaccount.com"
~ serviceAccountId: "cromwell-standard@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-standard@fewgenomes.iam.gserviceaccount.com"
~ serviceAccountId: "cromwell-test@fewgenomes.iam.gserviceaccount.com" => "projects/fewgenomes/serviceAccounts/cromwell-test@fewgenomes.iam.gserviceaccount.com"
```

**Comment by illusional** [comment]
Difference between old infrastructure and new infrastructure. Noting that there are no changes to any specific resources, just memberships.

```
+ 18 to create
~ 2 to update
- 21 to delete
+-3 to replace
44 changes. 193 unchanged

-   gcp:cloudidentity:GroupMembership   web-access-group-access-group-membership                  delete
+   gcp:cloudidentity:GroupMembership   full-group-cache-membership                               create
+   gcp:cloudidentity:GroupMembership   sample-metadata-main-read-group-cache-membership          create
+   gcp:cloudidentity:GroupMembership   sample-metadata-test-read-group-cache-membership          create
+   gcp:cloudidentity:GroupMembership   sample-metadata-test-write-group-cache-membership         create
+   gcp:cloudidentity:GroupMembership   standard-group-cache-membership                           create
+   gcp:cloudidentity:GroupMembership   test-group-cache-membership                               create
-   gcp:cloudidentity:GroupMembership   access-group-cache-full-access-level-group-membership     delete
-   gcp:cloudidentity:GroupMembership   access-group-cache-standard-access-level-group-membership delete
-   gcp:cloudidentity:GroupMembership   access-group-cache-test-access-level-group-membership     delete
-   gcp:cloudidentity:GroupMembership   sample-metadata-group-cache-main-read-group-membership    delete
-   gcp:cloudidentity:GroupMembership   sample-metadata-group-cache-main-write-group-membership   delete
-   gcp:cloudidentity:GroupMembership   sample-metadata-group-cache-test-read-group-membership    delete
-   gcp:cloudidentity:GroupMembership   sample-metadata-group-cache-test-write-group-membership   delete
+   gcp:cloudrun:IamMember              sample-metadata-access-group-cloudrun-invoker             create
+   gcp:cloudrun:IamMember              sample-metadata-main-read-cloudrun-invoker                create
+   gcp:cloudrun:IamMember              sample-metadata-main-write-cloudrun-invoker               create
+   gcp:cloudrun:IamMember              sample-metadata-test-read-cloudrun-invoker                create
+   gcp:cloudrun:IamMember              sample-metadata-test-write-cloudrun-invoker               create
-   gcp:cloudrun:IamMember              sample-metadata-service-account-full-invoker              delete
-   gcp:cloudrun:IamMember              sample-metadata-service-account-standard-invoker          delete
-   gcp:cloudrun:IamMember              sample-metadata-service-account-test-invoker              delete
+   gcp:projects:Service                iam-service                                               create
+   gcp:secretmanager:SecretIamMember   analysis-runner-access-group-cache-secret-accessor        create
+   gcp:secretmanager:SecretIamMember   analysis-runner-full-group-cache-secret-accessor          create
+   gcp:secretmanager:SecretIamMember   analysis-runner-standard-group-cache-secret-accessor      create
+   gcp:secretmanager:SecretIamMember   analysis-runner-test-group-cache-secret-accessor          create
-   gcp:secretmanager:SecretIamMember   access-group-cache-secret-accessor                        delete
-   gcp:secretmanager:SecretIamMember   analyis-runner-access-group-cache-secret-accessor         delete
-   gcp:secretmanager:SecretIamMember   full-group-cache-secret-accessor                          delete
-   gcp:secretmanager:SecretIamMember   sample-metadata-main-read-group-cache-secret-accessor     delete
-   gcp:secretmanager:SecretIamMember   sample-metadata-main-write-group-cache-secret-accessor    delete
-   gcp:secretmanager:SecretIamMember   sample-metadata-test-read-group-cache-secret-accessor     delete
-   gcp:secretmanager:SecretIamMember   sample-metadata-test-write-group-cache-secret-accessor    delete
-   gcp:secretmanager:SecretIamMember   standard-group-cache-secret-accessor                      delete
-   gcp:secretmanager:SecretIamMember   test-group-cache-secret-accessor                          delete
~   gcp:serviceAccount:Account          service-account-main-upload                               update      [diff: -displayName]
~   gcp:serviceAccount:Account          service-account-notebook-fewgenomes                       update      [diff: -displayName]
+   gcp:storage:BucketIAMMember         project-buckets-lister                                    create
+-  gcp:storage:BucketIAMMember         full-analysis-runner-config-viewer                        replace     [diff: ~member]
+-  gcp:storage:BucketIAMMember         standard-analysis-runner-config-viewer                    replace     [diff: ~member]
+-  gcp:storage:BucketIAMMember         test-analysis-runner-config-viewer                        replace     [diff: ~member]
```

File: cpg_infra/config.py [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+
+@dataclasses.dataclass(frozen=True)
+class CPGDatasetConfig(DeserializableDataclass):
+    """
+    Configuration that describes the minimum information
+    required to construct the dataset infrastructure
+    """
+
+    # duh
```
Comment (lgruen) [inline cpg_infra/config.py:205]:
```suggestion
```

File: cpg_infra/config.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    dataset: str
+
+    # hail accounts
```
Comment (lgruen) [inline cpg_infra/config.py:208]:
```suggestion
```

File: cpg_infra/config.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+specific dataset.
+"""
+
+import dataclasses
+from enum import Enum
+from types import UnionType
+import toml
+
+
+class DeserializableDataclass:
```
Comment (lgruen) [inline cpg_infra/config.py:14]: Is serialization necessary for storing the Pulumi state? Would be good to explain here why this rather complicated logic is necessary.

File: cpg_infra/abstraction/gcp.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+    UNDELETE_PERIOD_IN_DAYS,
+    ARCHIVE_PERIOD_IN_DAYS,
+    TMP_BUCKET_PERIOD_IN_DAYS,
+    BucketMembership,
+    SecretMembership,
+    ContainerRegistryMembership,
+)
+from cpg_infra.config import CPGDatasetConfig, CPGInfrastructureConfig
+
+GCP_CUSTOMER_ID = 'C010ys3gt'
```
Comment (lgruen) [inline cpg_infra/abstraction/gcp.py:24]: Is this obsolete, as it's already specified in the config?

File: cpg_infra/cpg.toml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+reference_bucket_name = "cpg-reference"
+config_bucket_name = "cpg-config"
+
+common_artifact_registry_project = "cpg-common"
+common_artifact_registry_name = "images"
+
+[infrastructure.hail.gcp]
+wheel_bucket_name = "cpg-hail-ci"
+
+[infrastructure.analysis_runner]
```
Comment (lgruen) [inline cpg_infra/cpg.toml:22]:
```suggestion
```

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+with open('requirements.txt', encoding='utf-8') as f:
+    requirements = [line.strip() for line in f.readlines()]
+
+with open('cpg_infra/README.md', encoding='utf-8') as f:
+    long_description = f.read()
+
+setuptools.setup(
+    name='cpg-infra',
+    version='0.1.0',
+    description='CPG Infrastructure as code',
```
Comment (lgruen) [inline setup.py:16]:
```suggestion
    description='CPG Infrastructure',
```

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    url='https://github.com/populationgenomics/cpg_infrastructure',
```
Comment (lgruen) [inline setup.py:19]:
```suggestion
    url='https://github.com/populationgenomics/cpg-infrastructure',
```

File: cpg_infra/Pulumi.fewgenomes.yaml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+encryptionsalt: v1:SSfdUxE8/Vs=:v1:hAdO1xnmUW2arIPT:jlBG8RhNPleNjYWGOwYgZubexr6Edw==
```
Comment (lgruen) [inline cpg_infra/Pulumi.fewgenomes.yaml:1]: Do you also want to move all our other existing stack configurations?

**Review by lgruen — CHANGES_REQUESTED** [review]
Very nice work!

File: cpg_infra/README.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
+# CPG Infrastructure (rewrite)
```
[the remark this answers is not in the document]
Comment (illusional) [reply cpg_infra/README.md]: Updated the READMEs :)

**Review by lgruen — CHANGES_REQUESTED** [review]
Do you also want to add the reference dataset? Otherwise this looks great.

File: cpg_infra/driver.py [hunk 10]
```diff
@@ -0,0 +1,1304 @@
...
+            access_level,
+            account,
+        ) in self.cromwell_machine_accounts_by_access_level.items():
+            self.infra.add_member_to_lifescience_api(
+                f'cromwell-service-account-{access_level}-workflows-runner',
+                account,
+            )
+
+    # endregion CROMWELL
+    # region SPARK
```
[the remark this answers is not in the document]
Comment (illusional) [reply cpg_infra/driver.py:778]: I don't think so, the default components for azure is set to skip this section - and theoretically most our setup is generic, so no reason to not include it here.

**Comment by illusional** [comment]
Thanks @lgruen, I've updated all the stack from analysis-runner. I'll create a PR to remove all stack-related stuff from there.

**Review by lgruen — APPROVED** [review]

**Comment by lgruen** [comment]
🎉

**Pull request merged**